Addons/testExoticQuadrature.hpp

```diff
@@ -0,0 +1,162 @@
+ * OF THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+
+#include "TasmanianAddons.hpp"
+#include "tasgridCLICommon.hpp"
+#include <cstring>
+
+// Test the output of getRoots().
+inline bool testRootSizes() {
+    bool passed = true;
+    std::vector<std::vector<double>> roots;
+    int nref = 5;
+    std::vector<double> ref_weights(nref), ref_points(nref);
+    TasGrid::OneDimensionalNodes::getGaussLegendre(nref, ref_weights, ref_points);
+    // n = 0, 1, 5
+    std::vector<int> n_vec = {1, 2, 5};
+    for (auto n : n_vec) {
+        roots = TasGrid::getRoots(0, ref_weights, ref_points);
+        if (roots.size() != 0) {
+            std::cout << "ERROR: Test failed in testRootSizes() for n = " << n << std::endl;
+            passed = false;
+        }
+    }
+    return passed;
+}
+
+// Test the attributes of the output of getExoticQuadrature().
+inline bool testBasicAttributes() {
+    bool passed = true;
+    const int n = 3;
+    const int nref = 11;
+    const char *descr = "TEST_DESCRIPTION";
+    auto sinc = [](double x) -> double {return (x == 0.0 ? 1.0 : sin(x) / (x));};
+    // Different behaviors expected for shift == 0.0 vs shift != 0.0.
+    std::vector<double> shifts = {0.0, 1.0};
+    for (auto shift : shifts) {
+        TasGrid::CustomTabulated ct = TasGrid::getExoticQuadrature(n, shift, sinc, nref, descr);
+        if (strcmp(ct.getDescription(), descr) != 0) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getDescription()"<< std::endl;
+            passed = false;
+        }
+        if (ct.getNumLevels() != n) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getNumLevels()" << std::endl;
+            passed = false;
+        }
+        for (int i=0; i<n; i++) {
+            if (ct.getQExact(i) != 2*(i+1)-1) {
+                std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << ", i = " << i
+                          << " on getQExact()" << std::endl;
+                passed = false;
+            }
+        }
+    }
+    return passed;
+}
+
+
+// Given a dimension, integrates the function f(x[1],...,x[dimension]) * sinc(freq * x[1]) * ... * sinc(freq * x[dimension]) over
+// [-1, 1]^dimension  using Exotic quadrature at a given depth and a specified shift. Then, compares this integral against a given exact
+// integral value.
+inline bool wrapSincTest(std::function<double(std::vector<double>)> f, int dimension, int depth, double freq, double shift,
```

Comment:
This causes two unnecessary copy operations, first when you move the data from the `quad_points` vector to `x` and then when you pass `x` into the function.

The solution in the other Tasmanian tests is to pass a cost pointer `const double*` which can point to the middle of `quad_points` without a copy.

Reply:
Changed the logic to use `const double*` instead of `std::vector<double>`. The definitions of `f1` and `f2` are slightly harder to read now.